### PR TITLE
Allow more complicated absolute paths in file_base

### LINKED ISSUE #26545: Error raised by Outputs/CSV object when relative path to `file_base` arg passes through the home directory `~`
## Bug Description

Error when attempting to use the CSV output object to write data to .csv file while the relative path from the current working directory to the `file_base` argument passes through the home directory `~`.

The problem may lie in the `find_first_of` function in FileOutput.C:

```
  // Check the file directory of file_base and create if needed
  // Check if _file_base is an absolute path
  std::string base;
  if (_file_base.find_first_of("/", 0) == 0)
    base = "./" + MooseUtils::relativepath(_file_base);
  else
    base = "./" + _file_base;
  base = base.substr(0, base.find_last_of('/'));

  if (_app.processor_id() == 0 && access(base.c_str(), W_OK) == -1)
  {
    // Directory does not exist. Loop through incremental directories and create as needed.
    std::vector<std::string> path_names;
    MooseUtils::tokenize(base, path_names);
    std::string inc_path = path_names[0];
    for (unsigned int i = 1; i < path_names.size(); ++i)
    {
      inc_path += '/' + path_names[i];
      if (access(inc_path.c_str(), W_OK) == -1)
        if (Utility::mkdir(inc_path.c_str()) == -1)
          mooseError("Could not create directory: " + inc_path + " for file base: " + _file_base);
    }
  }
```

## Steps to Reproduce

To reproduce, take some simple input file, e.g. one of the tutorials, and ensure there is a postprocessor variable and a CSV output object attempting to write it using the file_base argument, for example like this:
```
[Postprocessors]
  [max]
    type = ElementExtremeValue
    variable = temperature
  []
[]

[Outputs]
  exodus = true
  [write_to_csv]
    type = CSV
    show = 'max'
    file_base = '/home/user/absolute/path/to/desired/output'
  []
[]
```

Then if run from `~` for example, it will trigger an error like so:
```
*** ERROR ***
The following error occurred in the object "write_to_csv", of type "CSV".

Could not create directory: ./../ for file base: /home/user/absolute/path/to/desired/output
```

## Impact

Workaround discovered (see below) so this is non-blocking. However, it may  be useful for the CSV object to be able to handle this situation without the workaround in future.

Issue discovered when running Moose as a python [subprocess](https://docs.python.org/3/library/subprocess.html) in an optimisaton loop using [Ray-tune](https://docs.ray.io/en/latest/tune/index.html). Ray-tune changes the current working directory to one within `~/ray-results/` for each trial, leading to the error writing the csv file to a file base `~/something-else/` from this directory.

A workaround has been found by using the `cwd` kwarg to `subprocess.run()` and passing a suitable working directory.

### PR DESCRIPTION
**Pull request by lindsayad**
Closes #26545

### CODE REVIEW

**Comment by lindsayad** [comment]
@lukethehuman can you test this out?

File: framework/src/outputs/FileOutput.C [hunk 1]
```diff
@@ -148,12 +148,11 @@ FileOutput::setFileBaseInternal(const std::string & file_base)
 
   // Check the file directory of file_base and create if needed
   // Check if _file_base is an absolute path
-  std::string base;
-  if (_file_base.find_first_of("/", 0) == 0)
-    base = "./" + MooseUtils::relativepath(_file_base);
+  std::string base = _file_base.substr(0, _file_base.find_last_of('/'));
```
Comment (loganharbour) [inline framework/src/outputs/FileOutput.C:151]: is there not something we can do with `std::filesystem` that is smarter and easier to follow?

Comment (lindsayad) [reply framework/src/outputs/FileOutput.C]: moved to `std::filesystem`

**Review by loganharbour — COMMENTED** [review]
Looks good to me.

Do you have a particular odd combo that we should test with?

**Comment by lindsayad** [comment]
well the issue raised by @lukethehuman was with an absolute path. I can't really put an absolute path in file base since that's dependent on everyone's file system. I could create a test that converts a relative path to an absolute path and then passes that file base to `FileOutput`. But that is also exactly what the new code is doing... `std::filesystem::relative(std::filesystem::absolute` ... so I would be unit testing `std::filesystem`

**Review by loganharbour — APPROVED** [review]

**Comment by lindsayad** [comment]
failure is unrelated

**Pull request merged**